src/CodeGenerator/Generator/ApiGenerator.php

```diff
@@ -200,10 +200,135 @@ private function doGenerateResultClass(string $baseNamespace, string $className,
         }
 
         $this->resultClassAddProperties($baseNamespace, $root, $inputShape, $className, $class, $namespace);
+        // should be called After Properties injection
+        if ($operationName && null !== $pagination = $this->definition->getOperationPagination($operationName)) {
+            $this->generateOutputPagination($pagination, $className, $baseNamespace, $class);
+        }
 
         $this->fileWriter->write($namespace);
     }
 
+    private function generateOutputPagination(array $pagination, string $className, string $baseNamespace, ClassType $class)
+    {
+        if (empty($pagination['result_key'])) {
+            throw new \RuntimeException('This is not implemented yet');
+        }
+
+        $class->addImplement(\IteratorAggregate::class);
+
+        if (!\is_array($pagination['result_key'])) {
+            $resultShapeName = $this->definition->getShape($className)['members'][$pagination['result_key']]['shape'];
+            $resultShape = $this->definition->getShape($resultShapeName);
+
+            if ('list' !== $resultShape['type']) {
+                throw new \RuntimeException('Cannot generate a pagination for a non-iterable result');
+            }
+
+            $listShape = $this->definition->getShape($resultShape['member']['shape']);
+            if ('structure' === $listShape['type']) {
+                throw new \RuntimeException('This is not implemented yet');
+            }
+            $returnType = $this->toPhpType($listShape['type']);
+            $yield = "yield from \$this->{$pagination['result_key']};";
+        } else {
+            $mergedTypes = [];
+            foreach ($pagination['result_key'] as $key) {
+                $resultShapeName = $this->definition->getShape($className)['members'][$key]['shape'];
+                $resultShape = $this->definition->getShape($resultShapeName);
+                if ('list' !== $resultShape['type']) {
+                    throw new \RuntimeException('Cannot generate a pagination for a non-iterable result');
+                }
+
+                $listShape = $this->definition->getShape($resultShape['member']['shape']);
+                if ('structure' !== $listShape['type']) {
+                    throw new \RuntimeException('This is not implemented yet');
+                }
+                $type = $this->safeClassName($resultShape['member']['shape']);
+                $mergedTypes[] = $resultShape['member']['shape'];
+            }
+
+            $mergedClass = $this->generateMergeClass($mergedTypes, $baseNamespace);
+            $namespace = new PhpNamespace($baseNamespace);
+            $namespace->addUse($baseNamespace . $mergedClass);
+
+            $returnType = "{$mergedClass}";
```

Comment:
```suggestion
            $returnType = $mergedClass;
```

Isn't this the same?